lib/grammers-session/src/lib.rs

```diff
@@ -11,78 +11,110 @@
-                dcs: Vec::new(),
-                user: None,
-                state: None,
-            }),
-        }
+    pub fn save(&self) -> Result<(), SessionProviderError> {
+        self.provider.save()
     }
 
-    /// Load a previous session instance from a file,
-    /// creating one if it doesn't exist
-    pub fn load_file_or_create<P: AsRef<Path>>(path: P) -> io::Result<Self> {
-        let path = path.as_ref();
-        if !path.exists() {
-            File::create(path)?;
-            let session = Session::new();
-            session.save_to_file(path)?;
-            Ok(session)
-        } else {
-            Self::load_file(path)
-        }
+    pub fn load(&self) -> Result<(), SessionProviderError> {
+        self.provider.load()
     }
 
-    /// Load a previous session instance from a file.
-    pub fn load_file<P: AsRef<Path>>(path: P) -> io::Result<Self> {
-        let mut data = Vec::new();
-        File::open(path.as_ref())?.read_to_end(&mut data)?;
+    pub fn load_from_file_or_create<T: AsRef<Path> + 'static>(path: T) -> Result<Self, SessionProviderError> {
+        let fs_storage = FileSessionStorage::default_with_path(path)?;
+
+        match fs_storage.load() {
+            Ok(_) => (),
+            Err(SessionProviderError::InvalidFormat {source}) => {
+                let file_ref = fs_storage.get_file();
+                let mut file = file_ref.lock().unwrap();
+                let size = file.read_to_end(&mut vec![]).unwrap();
+                if size > 0 {
+                    return Err(SessionProviderError::InvalidFormat {source})
+                }
+            },
+            Err(e) => return Err(e)
+        };
+        let session = fs_storage.get_session();
+        let provider = Box::new(fs_storage);
 
-        Self::load(&data).map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))
+        Ok(
+            Self {
+                session,
+                provider,
+            }
+        )
     }
 
-    pub fn load(data: &[u8]) -> Result<Self, Error> {
-        Ok(Self {
-            session: Mutex::new(
-                enums::Session::from_bytes(data)
-                    .map_err(|e| match e {
-                        DeserializeError::UnexpectedEof => Error::MalformedData,
-                        DeserializeError::UnexpectedConstructor { .. } => Error::UnsupportedVersion,
-                    })?
-                    .into(),
-            ),
-        })
+    pub fn load_from_file<T: AsRef<Path> + 'static>(path: T) -> Result<Self, SessionProviderError> {
+        let fs_storage_options = FileSessionStorageOptions{
+            path: Box::new(path),
+            mode: OpenMode::Open,
+            session: Arc::new(Mutex::new(types::Session{
+                dcs: vec![],
+                user: None,
+                state: None,
+            }))
+        };
+
+        let fs_storage = FileSessionStorage::new(fs_storage_options)?;
+        fs_storage.load()?;
```

Comment:
This looks like a really strange way of using the API. I'd expect the load to return Self (and not need new).